Transfer ownership work with masterkey

sharidas opened the pull request:
This change helps users to get transfer-ownership
command work when masterkey is enabled as mode
of encryption.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
This change will help users to transfer files using transfer-ownership command when masterkey mode is enabled for encryption.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://github.com/owncloud/core/issues/27427

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
This change would help user tranfer files between users using transfer-ownership command when masterkey mode is enabled for encryption.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Below are the steps executed for testing:
`sujith@sujith-Inspiron-5567 ~/test/owncloud $ ./occ files:transfer-ownership --path transferred\ from\ user1\ on\ 20170608_115804/1 admin user1
Cannot load Xdebug - it was already loaded
Analysing files of admin ...
    1 [->--------------------------]
Check if master key enabled 1
    1 [============================]
Collecting all share information for files and folder of admin ...
    0 [>---------------------------]
Transferring files to user1/files/transferred from admin on 20170608_032359 ...
Restoring shares ...
    0 [>---------------------------]
sujith@sujith-Inspiron-5567 ~/test/owncloud $ ./occ files:transfer-ownership --path transferred\ from\ admin\ on\ 20170608_032359/1 user1 admin
Cannot load Xdebug - it was already loaded
Analysing files of user1 ...
    1 [============================]
Collecting all share information for files and folder of user1 ...
    0 [>---------------------------]
Transferring files to admin/files/transferred from user1 on 20170608_033618 ...
Restoring shares ...
    0 [>---------------------------]
sujith@sujith-Inspiron-5567 ~/test/owncloud $

This has also been tested by running the test transfer-ownership.feature`

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

sharidas committed:
Transfer ownership work with masterkey

This change helps users to get transfer-ownership
command work when masterkey is enabled as mode
of encryption.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented on lib/private/Files/Filesystem.php:
```diff
@@ -944,6 +944,9 @@ public static function getDirectoryContent($directory, $mimetype_filter = '') {
 	 */
 	public static function getPath($id) {
+		if (self::$defaultInstance === null) {
+			throw new NotFoundException("defaultInstance is null");
```
This is one solution I found to get rid of the  fatal error thrown by symfony. Any suggestions how to improve this?

PVince81 replied:
> thrown by symfony.

when exactly is this happening ? when running CI ?

sharidas replied:
> when exactly is this happening ? when running CI ?
When integration test is executed.

sharidas commented:
Attaching the integration-test result : 
[integration-test-result.txt](https://github.com/owncloud/core/files/1067142/integration-test-result.txt)

PVince81 commented on apps/encryption/lib/Crypto/Crypt.php:
```diff
@@ -480,7 +480,9 @@ public function symmetricDecryptFileContent($keyFileContents, $passPhrase, $ciph
 		$signature = $this->createSignature($data, $passPhrase);
 		if (!hash_equals($expectedSignature, $signature)) {
-			throw new HintException('Bad Signature', $this->l->t('Bad Signature'));
+			if (\OC::$server->getAppConfig()->getValue('encryption', 'useMasterKey', 0) === 0) {
```
As far as I know the master key should still use signatures to detect when encrypted files have been tampered with.

Last time there was another bug that led to "Bad signature" which was because something in the encryption code wasn't properly initialized. I suggest you verify that.

PVince81 replied:
You can read from here https://github.com/owncloud/core/issues/27261#issuecomment-282742094.
My solution was here https://github.com/owncloud/core/issues/27261#issuecomment-283680212 but will probably not apply. Still, having a look there and at my PR might give some potential clues.

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -194,6 +195,10 @@ function (FileInfo $fileInfo) use ($progress, $self) {
 					$this->allFiles[] = $fileInfo;
 					if ($fileInfo->isEncrypted()) {
+
+						if (\OC::$server->getAppConfig()->getValue('encryption', 'useMasterKey', 0) !== 0) {
```
why not add it to the array if it's encrypted with master key ? answer in a PHPDoc to de-confuse future devs

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -274,7 +279,10 @@ private function restoreShares(OutputInterface $output) {
+					try {
+						$this->shareManager->deleteShare($share);
+					} catch (\OCP\Files\NotFoundException $e) {
+					}
```
add error handling or a comment why we don't handle / ignore

sharidas commented:
Updated the PR by:
1) Reverted the change made in apps/encryption/lib/Crypto/Crypt.php
2) Added phpdoc for the exception in TransferOwnership.php
3) Tried to write the message to log file when the exception is called. Rather than making it empty
4) Added phpdoc explaining why the masterkey encryption doesn't need to be added to the array in TransferOwnership.php

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -274,7 +283,18 @@ private function restoreShares(OutputInterface $output) {
+						 * master key. The method getPath from \OC\Files\Filesystem tries
+						 * to call getPath using defaultInstance variable which
+						 * is null. Hence this causes our integration tests to break.
+						 */
```
Thanks for the explanation. I think we shouldn't catch and ignore an exception like this just because of integration tests.

I was fine with your other workaround in the other method when `instance` is null but here I think it's a bit too much.
Do you know which test fails ? Maybe there's a setupFS missing somewhere to initialize that default instance ?

From what I see the `defaultInstance` is created when calling `Filesystem::init()`.
Can you debug and compare why that one is properly set when running OCC manually and not when running integration tests ? There shouldn't be any difference...

sharidas replied:
I have updated the PR. This change is not required :( Hence I have removed the code in the latest PR. Here is the output of : OC_TEST_ENCRYPTION_MASTER_KEY_ENABLED=1 ./run.sh features/transfer-ownership.feature &> out.txt ( the out.txt : 
[out1.txt](https://github.com/owncloud/core/files/1091860/out1.txt) )

PVince81 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -194,6 +195,14 @@ function (FileInfo $fileInfo) use ($progress, $self) {
 					$progress->advance();
 					$this->allFiles[] = $fileInfo;
 					if ($fileInfo->isEncrypted()) {
+						if (\OC::$server->getAppConfig()->getValue('encryption', 'useMasterKey', 0) !== 0) {
```
Note that in general the transfer ownership app isn't supposed to know about specific encryption implementations.
Some people might implement their own custom encryption schemes, so would need a better abstraction.

But as we likely don't have time for better abstractions let's merge this.

PVince81 reviewed (approved):
:+1:

sharidas commented:
Updated the PR and the verified the transfer-ownership.feature on my machine: 
[log.txt](https://github.com/owncloud/core/files/1118863/log.txt)

sharidas commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -672,6 +672,16 @@ private function updateEncryptedVersion(Storage $sourceStorage, $sourceInternalP
+			 * version will be 1 and the expected version is at 2. And this results in Bad
+			 * Signature. This change helps to fix the issue.
+			 */
+			if ($encryptedVersion === 2) {
```
Not able to get any better solution. As per my understanding this version is causing "Bad Signature". The problem is when the expected signature is for version 2 and the signature generated is for version 1.

PVince81 replied:
Please try creating and overwriting a file multiple times before the transfer. You'll see that the value of `encrypted` in the database can be higher, so this hard-coded `2` likely will not work.

PVince81 replied:
I also suggest to add this as an additional integration test ^

sharidas commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -672,6 +672,13 @@ private function updateEncryptedVersion(Storage $sourceStorage, $sourceInternalP
+			 * Instead of the reusing old version, we stick with version 1. This helps
+			 * the data viewed in UI and helps resolve "Bad Signature".
+			 */
+			$encryptedVersion = 1;
```
Yes, that was right. When the data is updated in the file multiple times or file is edited multiple times, old logic fails. So if we make the version 1, then it looks like the data doesn't hit "Bad Signature".

PVince81 replied:
the "Bad signature" was only observed with transfer ownership or also other cases ?

cc @tomneedham

sharidas replied:
One of my observation is that whenever a new file is created the version is set to 0. So when transfer:ownership is called altogether a new file is being created. Which is resulting in version to set to zero: https://github.com/owncloud/core/blob/master/lib/private/Files/Storage/Wrapper/Encryption.php#L743. Whenever new file is created the version is set: https://github.com/owncloud/encryption/blob/master/lib/Crypto/Encryption.php#L192. Need to figure out how to use the sourceInternalPath's version assigned to targetInternalPath.

sharidas replied:
Did another experiment where I had 2 instances community edition(CE) and EE. Shared a folder of EE with CE. Now without my reset change, tried to create a file and moved to EE folder. It caused "Bad Signature". With my change enabled, created another file and moved to same EE folder. I was able to open the text file and read the contents. Moved the working file back to CE. No BadSignature was observed.

sharidas commented on tests/integration/features/transfer-ownership.feature:
```diff
@@ -12,7 +11,19 @@ Feature: transfer-ownership
 		Then Downloaded content when downloading file "/somefile.txt" with range "bytes=0-6" should be "This is"
 
-	@no_encryption
+	Scenario: transfering ownership of a file after updating the file
```
The intention behind this test is:
1) update the file 3 times
2) When the file is transferred it should be readable.

PVince81 replied:
I don't see the file updated three times. All you do is upload three separate pieces of the file. This doesn't increase the encryption version of the final file. You need to upload 3 times.
I suggest to not use chunked upload here for readability.

sharidas replied:
Ok. Will upload the same file 3 times. And then check the value in the file after download.

sharidas replied:
I have updated the feature by uploading same file 4 times. Before doing this, I had verified in the UI. When tried to update same file multiple times, the encrypted version was incremented.

PVince81 commented:
@sharidas so do I understand it correctly that when we transfer an encrypted file, whatever encryption version it has, we set it back to 1 or 2 and reencrypt for the new user ?

sharidas commented:
@PVince81 Yes. Any new file uploaded or lets say touched or created with content, the version will become 1. In the transfer-ownership we call https://github.com/owncloud/core/blob/b9850933a8fff67a4e4f0919ec66c238090feee7/apps/files/lib/Command/TransferOwnership.php#L263 rename which doesn't reset this. That's the reason the new user finds problematic to view the file.

SergioBertolinSG commented on tests/integration/features/transfer-ownership.feature:
```diff
@@ -12,7 +11,19 @@ Feature: transfer-ownership
+		And user "user0" uploads chunk file "2" of "3" with "BB" to "/PARENT/textfile0.txt"
+		And user "user0" uploads chunk file "3" of "3" with "CC" to "/PARENT/textfile0.txt"
+		When transfering ownership from "user0" to "user1"
+		And the command was successful
```
This is 'Then', the next ones are 'And'.

sharidas replied:
Updated. Now only one "Then" is there.

SergioBertolinSG commented on tests/integration/features/transfer-ownership.feature:
```diff
@@ -12,7 +11,19 @@ Feature: transfer-ownership
+		Then the command was successful
+		And As an "user1"
+		And using received transfer folder of "user1" as dav path
+		Then Downloaded content when downloading file "/PARENT/textfile0.txt" with range "bytes=0-5" should be "AABBCC"
```
This 'Then' has to be 'And'. (there can be only one then)

SergioBertolinSG commented:
:+1: to the integration tests.

PVince81 reviewed (approved):
:+1: if tests pass

PVince81 commented:
@sharidas strange failures, I suggest rebasing

sharidas commented:
Rebased the branch with current master.

sharidas commented:
It looks like the transfer-ownership.feature is run as ```./run.sh features/transfer-ownership.feature```. Which causes issue when run on the terminal. Where as if ran as ``` OC_TEST_ENCRYPTION_MASTER_KEY_ENABLED=1 ./run.sh features/transfer-ownership.feature``` doesn't cause any issue. Which also means need further investigation to get it working without setting OC_TEST_ENCRYPTION_MASTER_KEY_ENABLED.

PVince81 commented:
failing tests, please have a look

sharidas commented:
The test cases were passing without encryption: 
[log.txt](https://github.com/owncloud/core/files/1153407/log.txt) The file generated using command ```sujith@sujith-Inspiron-5567 ~/test/owncloud/tests/integration $ ./run.sh features/transfer-ownership.feature &> /tmp/log.txt```
and with encryption masterkey: 
[encrypt.txt](https://github.com/owncloud/core/files/1153426/encrypt.txt) The file generated using command : ```sujith@sujith-Inspiron-5567 ~/test/owncloud/tests/integration $ OC_TEST_ENCRYPTION_MASTER_KEY_ENABLED=1 ./run.sh features/transfer-ownership.feature &> /tmp/encrypt.txt```

sharidas commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -739,11 +744,17 @@ private function copyBetweenStorage(Storage $sourceStorage, $sourceInternalPath,
+				if($isRename) {
+					$absSourcePath = Filesystem::normalizePath($sourceStorage->getOwner($sourceInternalPath). '/' . $sourceInternalPath);
+					$absTargetPath = Filesystem::normalizePath($this->getFullPath($targetInternalPath));
+					\OC::$server->getSession()->set('renameFile',[$absSourcePath, $absTargetPath]);
```
One mechanism adopted to give a hint that rename action is taking place.

sharidas replied:
And associated changes here: https://github.com/owncloud/encryption/pull/8

sharidas commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -672,6 +672,12 @@ private function updateEncryptedVersion(Storage $sourceStorage, $sourceInternalP
+			 * Instead of the reusing old version, we stick with version 1. This helps
+			 * the data viewed in UI and helps resolve "Bad Signature".
+			 */
+			$cacheInformation['encryptedVersion'] += 1;
```
No reset. Instead increment the version.

sharidas commented:
This change works with transfer-ownership. The transfer of files between storage found problematic :(

PVince81 commented on lib/public/Encryption/IEncryptionModule.php:
```diff
@@ -58,13 +58,15 @@ public function getDisplayName();
 	 * @since 8.1.0
 	 */
-	public function begin($path, $user, $mode, array $header, array $accessList);
+	public function begin($path, $user, $mode, array $header, array $accessList, $renameFile);
```
since you are using a string instead of bool I suggest renaming this everywhere to `$sourceFileOfRename = null`.

sharidas commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -672,6 +673,12 @@ private function updateEncryptedVersion(Storage $sourceStorage, $sourceInternalP
+			 * Rename is a process of creating a new file. Here we try to use the
+			 * incremented version of source file, for the destination file.
+			 */
+			$encryptedVersion = $sourceStorage->getCache()->get($sourceInternalPath)['encryptedVersion'];
```
Modified this change. Trivial change. There was failure in couple of tests : apps/files_trashbin/tests/StorageTest.php and apps/files_trashbin/tests/TrashbinTest.php

PVince81 commented:
@sharidas please backport to stable10
